Draft Variable Sets API docs (#1930)
hashicorp/terraform-website: content/source/docs/cloud/api/variable-sets.html.md

---
layout: "cloud"
page_title: "Variable Sets - API Docs - Terraform Cloud and Terraform Enterprise"
---

# Variable Sets API

This set of APIs covers create, update, list, delete, and some specialized operations.

> Viewing Variable Sets requires access to the owning organization. Destructive actions require org admin status or workspace management permissions.

Reviewer: ```suggestion
Viewing Variable Sets requires access to the owning organization. Destructive actions require organization admin status or workspace management permissions.
```

Not a fan of shorthand, but if we do it elsewhere. 🤷‍♂️

## Create a Variable Set
`POST organizations/:organization_name/varsets`

Example Payload:
```json

> {:data=>

Reviewer: This appears to be Ruby, rather than JSON.

Author: d'oh

  {
    :type=>"varsets",
    :attributes=>{
      :name=>"varset-7f7b6783",
      :description=>"Full of vars and such for mass reuse",
      :is_global=>false
    },
    :relationships=>{
      :workspaces=>{
        !! Sending an empty workspaces list is an explicit declaration to unassign from all workspaces
        :data=>[
          {
            :id=>"ws-z6YvbWEYoE168kpq",
            :type=>"workspaces"
          },
          ...
        ]
      },
      :vars=>{
        :data=>[
          {
            :type=>"vars",
            :attributes=>{
              :key=>"c2e4612d993c18e42ef30405ea7d0e9ae",
              :value=>"8676328808c5bf56ac5c8c0def3b7071",
              :category=>"terraform"
            }
          },
          ...
        ]
      }
    }
  }
}
```

> rsponds with JSON representation of the save Variable Set (see list Variable Set section for example)

Reviewer: ```suggestion
responds with JSON representation of the save Variable Set (see list Variable Set section for example)
```

## Update a Variable Set
`PATCH varsets/:external_id`

> rsponds with JSON representation of the save Variable Set (see list Variable Set section for example)

Reviewer: Looks like a copy-pasta. 🍝

## Delete a Variable Set

> `PUT/PATCH varsets/:external_id`

Reviewer: `DELETE` I would think?

on success, responds with no content

## List Variable Set

> `GET organizations/:organization_name/varsets`

Reviewer: Can we differentiate these endpoints in the documentation? Should workspaces be under workspaces? Organizations under organizations?

---
layout: "cloud"
page_title: "Variable Sets - API Docs - Terraform Cloud and Terraform Enterprise"
---

# Variable Sets API

This set of APIs covers create, update, list, delete, and some specialized operations.

Viewing Variable Sets requires access to the owning organization. Destructive actions require organization admin status or workspace management permissions.

## Create a Variable Set
`POST organizations/:organization_name/varsets`

`data.name`                              | string         |         | The name of the Variable set
`data.description`                       | string         |         | A helpful blurb to contextualize the variable set
`data.is_global`                         | boolean        |         | When true, the variable set applies to all workspaces in the organization.

`data.relationships.workspaces`          | array          |         | Array of references to workspaces that the variable set should be assigned to
`data.relationships.vars`                | array          |         | Array of complete variable definitions that comprise the variable set.

> Example Payload:

Reviewer 2: ```suggestion
### Sample Payload
```

      is-global: false
    },
    relationships: {
      workspaces: {
        !! Sending an empty workspaces list is an explicit declaration to unassign from all workspaces
        data: [
          {
            id: "ws-z6YvbWEYoE168kpq",
            type: "workspaces"
          },
          ...
        ]
      },
      vars: {
        data: [
          {
            type: "vars",
            attributes: {
              key: "c2e4612d993c18e42ef30405ea7d0e9ae",
              value: "8676328808c5bf56ac5c8c0def3b7071",
              category: "terraform"
            }
          },
          ...
        ]
      }
    }
  }
}
```

responds with JSON representation of the save Variable Set (see list Variable Set section for example)

## Update a Variable Set
`PUT/PATCH varsets/:external_id`

rsponds with JSON representation of the saved Variable Set (see list Variable Set section for example)

## Delete a Variable Set
`DELETE varsets/:external_id`

on success, responds with no content

## List Variable Set
`GET organizations/:organization_name/varsets`

or

`GET workspaces/:workspace_external_id/varsets`

or

`GET varsets/:external_id` #same response sans pagination and listing

> example payload

Reviewer 2: ```suggestion
### Sample Payload
```

```json
{
  data: [
    {
      id: "varset-mio9UUFyFMjU33S4",
      type: "varsets",
      attributes:  {
         name: "varset-b7af6a77",
         description: "Full of vars and such for mass reuse",
         "is-global": false,
         "updated-at": "2021-10-29T17:15:56.722Z",
         "var-count":  5,
         "workspace-count": 2
      },
      relationships: {
        organization: {
         data: {id: "organization_1", type: "organizations"}
        },
        vars: {
          data: [
           {id: "var-abcd12345", type: "vars"},
           {id: "var-abcd12346", type: "vars"},
           {id: "var-abcd12347", type: "vars"},
           {id: "var-abcd12348", type: "vars"},
           {id: "var-abcd12349", type: "vars"}
          ]
        },
        workspaces: {
          data: [
           {id: "ws-abcd12345", type: "workspaces"},
           {id: "ws-abcd12346", type: "workspaces"}
          ]
        }
      }
    },
    {id: "varset-l1k3j41nf24jas",
      ... 
    },
    ...
  ],
  links: {
    self: "<page URL>",
    first: "<page URL>",
    prev: nil,
    next: nil,
    last: "<page URL>"
  }
}
```

> ## Relationships
> 
> ### Variables
> 
> ## Add Variable

Reviewer 2: These headers are confusing me :-) Typically we have relationships at the end of the  API docs, so I was surprised to see it here. Also, there doesn't seem to be anything under the variables level 3 header?

```suggestion
## Relationships

### Variables

## Add a Variable to a Variable Set
```

Reviewer 2: I'm still confused by this, I think!

`POST varsets/:varset_external_id/relationships/vars`

## Update a Variable in a Variable Set
`PATCH varsets/:varset_external_id/relationships/vars/:external_id`

> ```json

Reviewer 2: Is this a sample payload? If so, could we label it as such?

{
  data: {
    type: "vars",
    attributes: {
      key: "g6e45ae7564a17e81ef62fd1c7fa86138",
      value: "61e400d5ccffb3782f215344481e6c82",
      description: "cheeeese",
      sensitive: false,
      category: "terraform",
      hcl: false
    }
  }
}
```

> On success responds with data for saved variable

Reviewer 2: Can we give an example response?

Reviewer 2: ```suggestion
On success, this endpoint responds with the data for the updated variable.
```

## Delete a Variable in a Variable Set
`DELETEvarsets/:varset_external_id/relationships/vars/:external_id`

> on success, responds with deleted variable content

Reviewer 2: I think we need to spell out the parameters and give an example response.

Reviewer 2: ```suggestion
`DELETEvarsets/:varset_external_id/relationships/vars/:external_id`

On success, this endpoint responds with the content of the deleted variables.
```

> # List Variables

Reviewer 2: ```suggestion
## List Variables in a Variable Set
```

`GET varsets/:varset_external_id/relationships/vars`

> example response

Reviewer 2: ```suggestion
### Sample Response
```

```json
{
  data: [
    {
      id: "var-134r1k34nj5kjn",
      type: "vars",
      attributes: {
        key: "F115037558b045dd82da40b089e5db745",
        value: "1754288480dfd3060e2c37890422905f",
        sensitive: false,
        category: "terraform",
        hcl: false,
        "created-at": "2021-10-29T18:54:29.379Z",
        description: nil
      },
      relationships: {
        varset: {
          data: {
            id: "varset-992UMULdeDuebi1x",
            type: "varsets"},
          links: { related: "/api/v2/varsets/1" }
        }
      },
      links: { self: "/api/v2/vars/var-BEPU9NjPVCiCfrXj" }
    },
    ...
  ],
  links: {
    self: "<page URL>",
    first: "<page URL>",
    prev: nil,
    next: nil,
    last: "<page URL>"
  }
}
```

### Workspaces

## Assign Variable Set to Workspaces
`POST varsets/:varset_external_id/relationships/workspaces`

> payload

Reviewer 2: ```suggestion
### Sample Payload
```

```json
{
  data: [
    {
      type: "workspaces",
      id: "ws-YwfuBJZkdai4xj9w"
    },
    ...
  ]
}
```

> on success, responds with no content

Reviewer 2: ```suggestion
On success, this endpoint responds with no content.
```

> ## Unassign Variable Set from Workspaces

Reviewer 2: ```suggestion
## Remove a Variable Set from Workspaces
```

`DELETE varsets/:varset_external_id/relationships/workspaces`

> payload

Reviewer 2: ```suggestion
### Sample Payload
```

```json
{
  data: [
    {
      type: "workspaces",
      id: "ws-YwfuBJZkdai4xj9w"
    },
    ...
  ]
}
```

> on success, responds with no content

Reviewer 2: ```suggestion
On success, this endpoint responds with no content.
```

---
layout: "cloud"

> page_title: "Variable Sets - API Docs - Terraform Cloud and Terraform Enterprise"

Reviewer 2: ```suggestion
page_title: "Variable Sets - API Docs - Terraform Cloud and Terraform Enterprise"
description: "Variable sets allow you to reuse variables across multiple workspaces. Create, read, update, and delete variable sets with the Terraform Cloud API."
```

---

# Variable Sets API

A[variable set](/docs/cloud/workspaces/variables.html#variable-sets) is a resource that allows you to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to one or all workspaces.

You need [`read variables` permission](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to view the variables for a particular workspace and to view the variable sets in the owning organization. You need [read and write variables permissions](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to create and edit variable sets. 

> ## Create a Variable Set

Reviewer 2: One question I have is - what happens if they try to create a new global variable set that has a conflict? Do they get an error back? If so, we should probably tell users this up front.

Author: Good call! I'll add that to the docs.

`POST organizations/:organization_name/varsets`

| Parameter            | Description                                                                                                                                                              |
| -------------------- | ----------------------------------------------------------------------------------------------------- |
| `:organization_name` | The name of the organization the workspace belongs to.                                                |

### Request Body

Properties without a default value are required.

| Key path                               | Type           | Default | Description
-----------------------------------------|----------------|---------|------------

> `data.name`                              | string         |         | The name of the Variable set

Reviewer 2: ```suggestion
`data.name`                              | string         |         | The name of the variable set
```

Reviewer 2: are there any length constraints here that we should tell folks about? Or disallowed characters?

> `data.description`                       | string         | `""`      | A helpful blurb to contextualize the variable set

Reviewer 2: ```suggestion
`data.description`                       | string         | `""`      | Text displayed in the UI to contextualize the variable set and its purpose.
```

> `data.is_global`                         | boolean        | `false` | When true, the variable set applies to all workspaces in the organization.

Reviewer 2: ```suggestion
`data.is_global`                         | boolean        | `false` | When true, Terraform Cloud automatically applies the variable set to all current and future workspaces in the organization.
```

  https://app.terraform.io/api/v2/varsets/varset-kjkN545LH2Sfercv
```

### Sample Response

```json
{
  "data": {
    "id": "varset-kjkN545LH2Sfercv"
    "type": "varsets",
    "attributes": {
      "name": "MyVarset",
      "description": "Full of vars and such for mass reuse. Now global!",
      "is-global": true
    },
    "relationships": {
      "vars": {
        "data": [
          {
            "id": "var-Nh0doz0hzj9hrm34qq"
            "type": "vars",
            "attributes": {
              "key": "c2e4612d993c18e42ef30405ea7d0e9ae",
              "value": "8676328808c5bf56ac5c8c0def3b7071",
              "category": "terraform"
            }
          }
        ]
      }
    }
  }
}
```

## Delete a Variable Set
`DELETE varsets/:varset_id`

| Parameter            | Description         |
| -------------------- | --------------------|
| `:varset_id`         | The variable set ID |

### Sample Request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request DELETE \
  https://app.terraform.io/api/v2/varsets/varset-kjkN545LH2Sfercv
```

on success, responds with no content

## Show Variable Set

> Fetches details on a specific variable set

Reviewer 2: ```suggestion
Fetch details about the specified variable set.
```

      },
      "relationships": {
        "organization": {
          "data: {"id": "organization_1", "type": "organizations"}
        },
        "vars": {
          "data": [
           {"id": "var-abcd12345", "type": "vars"},
           {"id": "var-abcd12346", "type": "vars"},
           {"id": "var-abcd12347", "type": "vars"},
           {"id": "var-abcd12348", "type": "vars"},
           {"id": "var-abcd12349", "type": "vars"}
          ]
        },
        "workspaces": {
          "data": [
           {"id": "ws-abcd12345", "type": "workspaces"},
           {"id": "ws-abcd12346", "type": "workspaces"}
          ]
        }
      }
    },
    ...
  ],
  "links": {
    "self": "<page URL>",
    "first": "<page URL>",
    "prev": nil,
    "next": nil,
    "last": "<page URL>"
  }
}
```

## Relationships

### Variables

## Add Variable

`POST varsets/:varset_external_id/relationships/vars`

| Parameter            | Description                      |
| -------------------- | ---------------------------------|
| `:varset_id`         | The variable set ID              |

### Request Body

This POST endpoint requires a JSON object with the following properties as a request payload.

> Properties without a default value are required.

Reviewer 2: Do we need to do this for every one of these that needs a JSON payload? I only see this language here, but I think we probably need to have it for all of the areas, right?

Author: Yeah, I definitely missed this bit in the copy/pasting.

    "attributes": {
      "key": "g6e45ae7564a17e81ef62fd1c7fa86138",
      "value": "61e400d5ccffb3782f215344481e6c82",
      "description": "cheeeese",
      "sensitive": false,
      "category": "terraform",
      "hcl": false
    }
  }
}
```

## Update a Variable in a Variable Set

`PATCH varsets/:varset_id/relationships/vars/:var_id`

| Parameter            | Description                      |
| -------------------- | ---------------------------------|
| `:varset_id`         | The variable set ID              |
| `:var_id`            | The ID of the variable to delete |

### Sample Payload

```json
{
  "data": {
    "type": "vars",
    "attributes": {
      "key": "g6e45ae7564a17e81ef62fd1c7fa86138",
      "value": "61e400d5ccffb3782f215344481e6c82",
      "description": "new cheeeese",
      "sensitive": false,
      "category": "terraform",
      "hcl": false
    }
  }
}
```

### Sample Request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request PATCH \
  --data @payload.json \
  https://app.terraform.io/api/v2/varsets/varset-4q8f7H0NHG733bBH/relationships/vars/var-EavQ1LztoRTQHSNT
```

> ### Sample Respone

Reviewer 2: ```suggestion
### Sample Response
```

```

on success, responds with no content

## List Variables in a Variable Set

`GET varsets/:varset_id/relationships/vars`

| Parameter            | Description         |
| -------------------- | --------------------|
| `:varset_id`         | The variable set ID |

### Sample Response

```json
{
  "data": [
    {
      "id": "var-134r1k34nj5kjn",
      "type": "vars",
      "attributes": {
        "key": "F115037558b045dd82da40b089e5db745",
        "value": "1754288480dfd3060e2c37890422905f",
        "sensitive": false,
        "category": "terraform",
        "hcl": false,
        "created-at": "2021-10-29T18:54:29.379Z",
        "description": nil
      },
      "relationships": {
        "varset": {
          "data": {
            "id": "varset-992UMULdeDuebi1x",
            "type": "varsets"
          },
          "links": { "related": "/api/v2/varsets/1" }
        }
      },
      "links": { "self": "/api/v2/vars/var-BEPU9NjPVCiCfrXj" }
    },
  ],
  "links": {
    "self": "<page URL>",
    "first": "<page URL>",
    "prev": nil,
    "next": nil,
    "last": "<page URL>"
  }
}
```

### Workspaces

> ## Assign Variable Set to Workspaces

Reviewer 2: In the docs, we're referring to this as "applying" the variable set, rather than assigning it
```suggestion
## Apply Variable Set to Workspaces
```

---
layout: "cloud"
page_title: "Variable Sets - API Docs - Terraform Cloud and Terraform Enterprise"
description: "Variable sets allow you to reuse variables across multiple workspaces. Create, read, update, and delete variable sets with the Terraform Cloud API."
---

# Variable Sets API

> A[variable set](/docs/cloud/workspaces/variables.html#variable-sets) is a resource that allows you to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to one or all workspaces.

Reviewer 2: ```suggestion
A [variable set](/docs/cloud/workspaces/variables.html#variable-sets) is a resource that allows you to reuse the same variables across multiple workspaces. For example, you could define a variable set of provider credentials and automatically apply it to one or all workspaces.
```

You need [`read variables` permission](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to view the variables for a particular workspace and to view the variable sets in the owning organization. You need [read and write variables permissions](/docs/cloud/users-teams-organizations/permissions.html#general-workspace-permissions) to create and edit variable sets. 

## Create a Variable Set

`POST organizations/:organization_name/varsets`

| Parameter            | Description                                                                                                                                                              |
| -------------------- | ----------------------------------------------------------------------------------------------------- |
| `:organization_name` | The name of the organization the workspace belongs to.                                                |

### Request Body

Properties without a default value are required.

| Key path                               | Type           | Default | Description
-----------------------------------------|----------------|---------|------------

> `data.name`                              | string         |         | The name of the variable set

Reviewer 2: just checking - any restrictions on what characters they can use/length?

```suggestion
`data.name`                              | string         |         | The name of the variable set.
```

`data.description`                       | string         | `""`      | Text displayed in the UI to contextualize the variable set and its purpose.
`data.is_global`                         | boolean        | `false` | When true, Terraform Cloud automatically applies the variable set to all current and future workspaces in the organization.
`data.relationships.workspaces`          | array          | []      | Array of references to workspaces that the variable set should be assigned to. Sending an empty array clears all workspace assignments.
`data.relationships.vars`                | array          | []      | Array of complete variable definitions that comprise the variable set.

> Trying to save 2 variables of the same name and category among any global variable sets will result in a 422 response.

Reviewer 2: ```suggestion
Terraform Cloud does not allow different global variable sets to contain conflicting variables with the same name and type. You will receive a 422 response if you try to create a global variable set that contains conflicting variables.
```

  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  --data @payload.json \
  https://app.terraform.io/api/v2/organziations/my-organization/varsets
```

### Sample Response

```json
{
  "data": {
    "id": "varset-kjkN545LH2Sfercv"
    "type": "varsets",
    "attributes": {
      "name": "MyVarset",
      "description": "Full of vars and such for mass reuse",
      "is-global": false
    },
    "relationships": {
      "workspaces": {
        "data": [
          {
            "id": "ws-z6YvbWEYoE168kpq",
            "type": "workspaces"
          }
        ]
      },
      "vars": {
        "data": [
          {
            "id": "var-Nh0doz0hzj9hrm34qq"
            "type": "vars",
            "attributes": {
              "key": "c2e4612d993c18e42ef30405ea7d0e9ae",
              "value": "8676328808c5bf56ac5c8c0def3b7071",
              "category": "terraform"
            }
          }
        ]
      }
    }
  }
}
```

## Update a Variable Set

`PUT/PATCH varsets/:varset_id`

| Parameter            | Description         |
| -------------------- | --------------------|
| `:varset_id`         | The variable set ID |

> Trying to save 2 variables of the same name and category among any global variable sets will result in a 422 response.

Reviewer 2: ```suggestion
Terraform Cloud does not allow global variable sets to contain conflicting variables with the same name and type. You will receive a 422 response if you try to create a global variable set that contains conflicting variables.
```

```

### Sample Response

```json
{
  "data": {
    "id": "varset-kjkN545LH2Sfercv"
    "type": "varsets",
    "attributes": {
      "name": "MyVarset",
      "description": "Full of vars and such for mass reuse. Now global!",
      "is-global": true
    },
    "relationships": {
      "vars": {
        "data": [
          {
            "id": "var-Nh0doz0hzj9hrm34qq"
            "type": "vars",
            "attributes": {
              "key": "c2e4612d993c18e42ef30405ea7d0e9ae",
              "value": "8676328808c5bf56ac5c8c0def3b7071",
              "category": "terraform"
            }
          }
        ]
      }
    }
  }
}
```

## Delete a Variable Set
`DELETE varsets/:varset_id`

| Parameter            | Description         |
| -------------------- | --------------------|
| `:varset_id`         | The variable set ID |

### Sample Request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request DELETE \
  https://app.terraform.io/api/v2/varsets/varset-kjkN545LH2Sfercv
```

On success, this endpoint responds with no content.

## Show Variable Set

Fetch details about the specified variable set.

> `GET varsets/:varset_id` #same response sans pagination and listing

Reviewer 2: the same response is confusing to me - what does that mean?

```suggestion
`GET varsets/:varset_id` #same response sans pagination and listing
```

Author: Oh good catch. Removed.

      "type": "varsets",
      "attributes":  {
         "name": "varset-b7af6a77",
         "description": "Full of vars and such for mass reuse",
         "is-global": false,
         "updated-at": "2021-10-29T17:15:56.722Z",
         "var-count":  5,
         "workspace-count": 2
      },
      "relationships": {
        "organization": {
          "data: {"id": "organization_1", "type": "organizations"}
        },
        "vars": {
          "data": [
           {"id": "var-abcd12345", "type": "vars"},
           {"id": "var-abcd12346", "type": "vars"},
           {"id": "var-abcd12347", "type": "vars"},
           {"id": "var-abcd12348", "type": "vars"},
           {"id": "var-abcd12349", "type": "vars"}
          ]
        },
        "workspaces": {
          "data": [
           {"id": "ws-abcd12345", "type": "workspaces"},
           {"id": "ws-abcd12346", "type": "workspaces"}
          ]
        }
      }
    }
  ],
  "links": {
    "self": "app.terrafor.io/app/my_organization/varsets",
    "first": "app.terrafor.io/app/my_organization/varsets?page=1",
    "prev": null,
    "next": null,
    "last": "app.terrafor.io/app/my_organization/varsets?page=1"
  }
}
```

> ### Variable Relationships

Reviewer 2: should there be something under this heading? It's a level three :-)

| Parameter            | Description                      |
| -------------------- | ---------------------------------|
| `:varset_id`         | The variable set ID              |

### Request Body

This POST endpoint requires a JSON object with the following properties as a request payload.

Properties without a default value are required.

Key path                                 | Type   | Default | Description
-----------------------------------------|--------|---------|------------
`data.type`                              | string |         | Must be `"vars"`.
`data.attributes.key`                    | string |         | The name of the variable.
`data.attributes.value`                  | string | `""`    | The value of the variable.
`data.attributes.description`            | string |         | The description of the variable.
`data.attributes.category`               | string |         | Whether this is a Terraform or environment variable. Valid values are `"terraform"` or `"env"`.
`data.attributes.hcl`                    | bool   | `false` | Whether to evaluate the value of the variable as a string of HCL code. Has no effect for environment variables.

> `data.attributes.sensitive`              | bool   | `false` | Whether the value is sensitive. If true then the variable is written once and not visible thereafter.

Reviewer 2: Docs say:

"Terraform runs receive the full text of [sensitive](#sensitive-values) Terraform variables and may print the value in logs and state files if the configuration sends the value to an output or a resource parameter. Sentinel mocks downloaded from runs will also contain the sensitive values of Terraform variables."

Maybe we shuld change this to read "If true, variable is not visible in the UI."

> Trying to save 2 variables of the same name and category among any global variable sets will result in a 422 response.

Reviewer 2: ```suggestion
Terraform Cloud does not allow different global variable sets to contain conflicting variables with the same name and type. You will receive a 422 response if you try to add a conflicting variable to a global variable set.
```

    "attributes": {
      "key": "g6e45ae7564a17e81ef62fd1c7fa86138",
      "value": "61e400d5ccffb3782f215344481e6c82",
      "description": "cheeeese",
      "sensitive": false,
      "category": "terraform",
      "hcl": false
    }
  }
}
```

### Sample Request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  --data @payload.json \
  https://app.terraform.io/api/v2/varsets/varset-4q8f7H0NHG733bBH/relationships/vars
```

### Sample Respone

```json
{
  "data": {
    "id":"var-EavQ1LztoRTQHSNT"
    "type": "vars",
    "attributes": {
      "key": "g6e45ae7564a17e81ef62fd1c7fa86138",
      "value": "61e400d5ccffb3782f215344481e6c82",
      "description": "cheeeese",
      "sensitive": false,
      "category": "terraform",
      "hcl": false
    }
  }
}
```

## Update a Variable in a Variable Set

`PATCH varsets/:varset_id/relationships/vars/:var_id`

| Parameter            | Description                      |
| -------------------- | ---------------------------------|
| `:varset_id`         | The variable set ID              |
| `:var_id`            | The ID of the variable to delete |

> Trying to save 2 variables of the same name and category among any global variable sets will result in a 422 response.

Reviewer 2: ```suggestion
Terraform Cloud does not allow different global variable sets to contain conflicting variables with the same name and type. You will receive a 422 response if you try to add a conflicting variable to a global variable set.
```

```json
{
  "data": [
    {
      "id": "var-134r1k34nj5kjn",
      "type": "vars",
      "attributes": {
        "key": "F115037558b045dd82da40b089e5db745",
        "value": "1754288480dfd3060e2c37890422905f",
        "sensitive": false,
        "category": "terraform",
        "hcl": false,
        "created-at": "2021-10-29T18:54:29.379Z",
        "description": ""
      },
      "relationships": {
        "varset": {
          "data": {
            "id": "varset-992UMULdeDuebi1x",
            "type": "varsets"
          },
          "links": { "related": "/api/v2/varsets/1" }
        }
      },
      "links": { "self": "/api/v2/vars/var-BEPU9NjPVCiCfrXj" }
    },
  ],
  "links": {
    "self": "app.terrafor.io/app/varsets/varset-992UMULdeDuebi1x/vars",
    "first": "app.terrafor.io/app/varsets/varset-992UMULdeDuebi1x/vars?page=1",
    "prev": null,
    "next": null,
    "last": "app.terrafor.io/app/varsets/varset-992UMULdeDuebi1x/vars?page=1"
  }
}
```

### Workspaces Relationships

## Apply Variable Set to Workspaces

`POST varsets/:varset_id/relationships/workspaces`

| Parameter            | Description         |
| -------------------- | --------------------|

> | `:varset_id`         | The variable set ID |

Reviewer 2: Can we do multiple at a time, or is it just one workspace per request?

Author: Yep! updating to reflect that.

### Sample Payload

```json
{
  "data": [
    {
      "type": "workspaces",
      "id": "ws-YwfuBJZkdai4xj9w"
    }
  ]
}
```

### Sample Request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  https://app.terraform.io/api/v2/varsets/varset-kjkN545LH2Sfercv/relationships/workspaces
```

On success, this endpoint responds with no content.

> ## Remove a Variable Set from Workspaces

Reviewer 2: This makes it seem like I could do multiple workspaces - but there's just one ID. Should we clarify? CAN you remove from more than one at once, or is it just one at a time?
